# Customer Lifetime Value Anaylsis

In marketing Customer Lifetime Value (CLTV or CLV) is a prognostication of the net profit contributed to the whole future relationship with a customer. Or in other words, it is the monetary value that represents the amount of revenue or profits a customer will bring to the company over a period of the relationship.  And CLV will help the business understand its customers better in following ways.
•	Helps engagement with the customer appropriately
•	Helps in customer retention strategies
•	Can identify the most profitable customers and their segmentation
•	Can enhance customer acquisition strategies

<b> How to Calculate Customer Lifetime Value </b>
There are so many different ways to calculate Customer Lifetime Value. For this project we are using the following formula:

<div class="alert alert-block alert-info">
<b>Formula:</b> 
    
>$CLTV = ((Average Order Value * Purchase Frequency)/Churn Rate)* Profit Margin $ <br>

>$Customer Value = Average Order Value * Purchase Frequency $ <br>

</div>

Where:<br>
<b>Average Order Value(AOV)</b> is the ratio of Total revenue and the total number of orders. It is the Average amount that a customer spends on an order. <br>

<div class="alert alert-block alert-info">
<b>Formula:</b> 
    
>$AOV= Total Revenue/Total Number of Orders$ <br>
    
</div>

<b> Purchase Frequency(PF)</b> is the ration of the total number of orders and the total customers<br>
<div class="alert alert-block alert-info">
<b>Formula:</b> 
    
>$PF = Total Number of Order / Total number of Customers$<br>

</div>

<b>Churn Rate</b> is the percentage of customers who have not ordered again.<br>
<div class="alert alert-block alert-info">
<b>Formula:</b> 
    
>$Churn Rate = 1 - Repeat Rate$
    
</div>

<b>Customer Lifetime</b> is the period when the customer is ordering continuously. 
<div class="alert alert-block alert-info">
<b>Formula:</b> 
    
>$Customer Lifetime = 1/Churn Rate$
    
</div>
    
<b> Repeat Rate</b> is the ratio of number of customers with more than one order to the number of unique customers. For example id you have 5 customer bought in a month and out of those 5, 2 come back, your repeat rate is 40%<br>









## CLTV calcuation using Python

### Importing required Library 

In [2]:
import pandas as pd
import seaborn as sns

import matplotlib.pyplot as plt
import numpy as np
import datetime as dt

import missingno as msno
from textwrap import wrap

import os

<b> Loading Dataset </b> <br>

Before loading dataset, using OS library to reach the directory where file is save. And then using Panda Read Function to load the dataset

In [3]:
os.chdir('C:\\Users\\Dell\\OneDrive\Desktop\\Retention project file')
         

In [4]:
transaction_df = pd.read_excel('transaction.xlsx')

In [5]:
transaction_df.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [6]:
transaction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

<b> Taking Care of Missing Values</b> <br>

There was no way to find or retrieve the values of the blank fields or null values. So droping the rows using Dropna function

In [10]:
filtered_data = transaction_df.dropna(subset=['list_price', 'standard_cost'])

In [11]:
filtered_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19803 non-null  int64         
 1   product_id               19803 non-null  int64         
 2   customer_id              19803 non-null  int64         
 3   transaction_date         19803 non-null  datetime64[ns]
 4   online_order             19445 non-null  float64       
 5   order_status             19803 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               19803 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [12]:
filtered_data.describe()

transaction_id    product_id   customer_id  online_order    list_price  \
count    19803.000000  19803.000000  19803.000000  19445.000000  19803.000000   
mean      9993.979650     45.815937   1739.294955      0.500849   1107.997866   
std       5776.499961     30.569841   1011.946353      0.500012    582.770460   
min          1.000000      0.000000      1.000000      0.000000     12.010000   
25%       4988.500000     18.000000    858.000000      0.000000    575.270000   
50%       9995.000000     45.000000   1738.000000      1.000000   1163.890000   
75%      15000.500000     72.000000   2615.000000      1.000000   1635.300000   
max      20000.000000    100.000000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

<b>Flitering the required Columns</b><br> 

><b> Transaction_id </b> will help you count the frequency of the transaction performed <br>
><b> Customer_id </b> will help you find unique customers <br>
><b> Transaction_date</b> will help you find the number of days customer stays with the product<br>
><b> List_price</b> will provide you the selling price <br>
><b> Standard_cost</b> will provide the cost price<br>

In [13]:
filtered_data = filtered_data[['transaction_id','customer_id','transaction_date','list_price', 'standard_cost' ]]

In [14]:
filtered_data.head()

transaction_id  customer_id transaction_date  list_price  standard_cost
0               1         2950       2017-02-25       71.49          53.62
1               2         3120       2017-05-21     2091.47         388.92
2               3          402       2017-10-16     1793.43         248.82
3               4         3135       2017-08-31     1198.46         381.10
4               5          787       2017-10-01     1765.30         709.48

Here we are going to group by the data on Customer_id column and will perform following aggregation on rest of the coulmns.
*  Calculating number of days between recent purchase date  and last purchase date
*  Number of orders
*  Calculate sum of List price
*  Calculate sum of Standard price

In [16]:
transaction_group = filtered_data.groupby('customer_id').agg({'transaction_date' : lambda date : (date.max()-date.min()).days,
                                                             'transaction_id' : lambda num: len(num),
                                                             'list_price' : lambda price: price.sum(),
                                                             'standard_cost' : lambda cost: cost.sum()})

In [17]:
transaction_group.head()

transaction_date  transaction_id  list_price  standard_cost
customer_id                                                             
1                         352              11     9084.45        6066.36
2                         112               3     4149.07        1922.81
3                         208               8     9888.23        6525.42
4                          76               2     1047.72         827.15
5                         286               6     5903.20        3508.26

<b> Renaming the Column</b> 

In [20]:
transaction_group.columns = ['num_days','num_transactions','selling_price', 'cost']

In [21]:
transaction_group.head()

num_days  num_transactions  selling_price     cost
customer_id                                                    
1                 352                11        9084.45  6066.36
2                 112                 3        4149.07  1922.81
3                 208                 8        9888.23  6525.42
4                  76                 2        1047.72   827.15
5                 286                 6        5903.20  3508.26

### Calculate CLTV using Formula

<div class="alert alert-block alert-info">
<b>Formula:</b> 
    
>$CLTV = ((Average Order Value * Purchase Frequency)/Churn Rate)* Profit Margin $ <br>

>$Customer Value = Average Order Value * Purchase Frequency $ <br>

</div>

<b> 1. Calculating Average Order Value </b>

In [22]:
transaction_group['Avg_order_value'] = transaction_group['selling_price']/transaction_group['num_transactions']

<b> 2. Calculating Purchase Frequency </b> 

In [24]:
Purchase_freq = sum(transaction_group.num_transactions >1)/transaction_group.shape[0]

<b> 3. Calculating Repeat Rate and Churn Rate </b> 

In [25]:
Repeat_rate = transaction_group[transaction_group.num_transactions >1].shape[0]/transaction_group.shape[0]

In [26]:
churn_rate = 1-Repeat_rate

<b> Purchase Frequency, Repeat Rate, Churn Rate</b>

In [27]:
print(Purchase_freq, Repeat_rate, churn_rate)

0.9842587292501431 0.9842587292501431 0.015741270749856873


<b> 4. Calculating Profit Margin </b>

In [28]:
transaction_group['Profit_margin'] = transaction_group['selling_price'] - transaction_group['cost']

In [29]:
transaction_group.head()

num_days  num_transactions  selling_price     cost  \
customer_id                                                       
1                 352                11        9084.45  6066.36   
2                 112                 3        4149.07  1922.81   
3                 208                 8        9888.23  6525.42   
4                  76                 2        1047.72   827.15   
5                 286                 6        5903.20  3508.26   

             Avg_order_value  Profit_margin  
customer_id                                  
1                 825.859091        3018.09  
2                1383.023333        2226.26  
3                1236.028750        3362.81  
4                 523.860000         220.57  
5                 983.866667        2394.94

<b> 5. Calculating Customer Lifetime value 

In [30]:
#Customer Value
transaction_group['CLV'] = (transaction_group['Avg_order_value']* Purchase_freq)/churn_rate

In [71]:
#Customer Lifetime Value
transaction_group['Customer_lifetime_value']= transaction_group['CLV']* transaction_group['Profit_margin']

In [32]:
transaction_group.head()

num_days  num_transactions  selling_price     cost  \
customer_id                                                       
1                 352                11        9084.45  6066.36   
2                 112                 3        4149.07  1922.81   
3                 208                 8        9888.23  6525.42   
4                  76                 2        1047.72   827.15   
5                 286                 6        5903.20  3508.26   

             Avg_order_value  Profit_margin           CLV  \
customer_id                                                 
1                 825.859091        3018.09  51638.716612   
2                1383.023333        2226.26  86476.677152   
3                1236.028750        3362.81  77285.506750   
4                 523.860000         220.57  32755.537091   
5                 983.866667        2394.94  61518.499394   

             Customer_lifetime_value  
customer_id                           
1                       1.558503e+08  
2                       1.925196e+08  
3                       2.598965e+08  
4                       7.224889e+06  
5                       1.473331e+08

### Prediction Model for CLTV

Now we are going to built the prediction model for CLTV using Liner Regression Model

For this we are going to use loaded data and our filtered data. 

In [33]:
# prediction model

filtered_data.head()

transaction_id  customer_id transaction_date  list_price  standard_cost
0               1         2950       2017-02-25       71.49          53.62
1               2         3120       2017-05-21     2091.47         388.92
2               3          402       2017-10-16     1793.43         248.82
3               4         3135       2017-08-31     1198.46         381.10
4               5          787       2017-10-01     1765.30         709.48

Extrating month and year from the transaction date

In [34]:
filtered_data['month_yr']=filtered_data['transaction_date'].apply(lambda x: x.strftime('%b-%Y'))

C:\Users\Dell\AppData\Local\Temp\ipykernel_33572\1763148451.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data['month_yr']=filtered_data['transaction_date'].apply(lambda x: x.strftime('%b-%Y'))


In [35]:
filtered_data.head()

transaction_id  customer_id transaction_date  list_price  standard_cost  \
0               1         2950       2017-02-25       71.49          53.62   
1               2         3120       2017-05-21     2091.47         388.92   
2               3          402       2017-10-16     1793.43         248.82   
3               4         3135       2017-08-31     1198.46         381.10   
4               5          787       2017-10-01     1765.30         709.48   

   month_yr  
0  Feb-2017  
1  May-2017  
2  Oct-2017  
3  Aug-2017  
4  Oct-2017

Creating Pivot Table which takes the columns as input, and groups the entries into 2 dimensional table and will provide multi dimensional summarization of the data.

In [38]:
# creating Pivot Tablea

sale=filtered_data.pivot_table(index=['customer_id'], columns = ['month_yr'], values='list_price', aggfunc='sum', fill_value=0).reset_index()

In [39]:
sale.head()

month_yr  customer_id  Apr-2017  Aug-2017  Dec-2017  Feb-2017  Jan-2017  \
0                   1   1577.53      0.00   1209.08     71.49     360.4   
1                   2      0.00   1992.93      0.00      0.00       0.0   
2                   3   3134.50      0.00      0.00   1311.44       0.0   
3                   4    569.56      0.00      0.00      0.00       0.0   
4                   5   1163.89    774.53   1812.75      0.00       0.0   

month_yr  Jul-2017  Jun-2017  Mar-2017  May-2017  Nov-2017  Oct-2017  Sep-2017  
0              0.0    642.70   1963.56   3259.69       0.0       0.0      0.00  
1              0.0    752.64      0.00   1403.50       0.0       0.0      0.00  
2              0.0    363.01   1148.64      0.00       0.0       0.0   3930.64  
3              0.0    478.16      0.00      0.00       0.0       0.0      0.00  
4              0.0      0.00    688.63   1463.40       0.0       0.0      0.00

Summing up all the months Sales

In [41]:
sale['CLV']= sale.iloc[:,2:].sum(axis=1)

In [42]:
sale.head()

month_yr  customer_id  Apr-2017  Aug-2017  Dec-2017  Feb-2017  Jan-2017  \
0                   1   1577.53      0.00   1209.08     71.49     360.4   
1                   2      0.00   1992.93      0.00      0.00       0.0   
2                   3   3134.50      0.00      0.00   1311.44       0.0   
3                   4    569.56      0.00      0.00      0.00       0.0   
4                   5   1163.89    774.53   1812.75      0.00       0.0   

month_yr  Jul-2017  Jun-2017  Mar-2017  May-2017  Nov-2017  Oct-2017  \
0              0.0    642.70   1963.56   3259.69       0.0       0.0   
1              0.0    752.64      0.00   1403.50       0.0       0.0   
2              0.0    363.01   1148.64      0.00       0.0       0.0   
3              0.0    478.16      0.00      0.00       0.0       0.0   
4              0.0      0.00    688.63   1463.40       0.0       0.0   

month_yr  Sep-2017      CLV  
0             0.00  7506.92  
1             0.00  4149.07  
2          3930.64  6753.73  
3             0.00   478.16  
4             0.00  4739.31

In [43]:
sale.columns

Index(['customer_id', 'Apr-2017', 'Aug-2017', 'Dec-2017', 'Feb-2017',
       'Jan-2017', 'Jul-2017', 'Jun-2017', 'Mar-2017', 'May-2017', 'Nov-2017',
       'Oct-2017', 'Sep-2017', 'CLV'],
      dtype='object', name='month_yr')

Now we are going to divide the data for Training and Testing the module.

#### Selecting Features for Trainnig and Testing

Here we are going to divide the columns into X and Y Variable where X is the last six months data as independent variable and Y as the Total sales as dependent variable.

In [50]:
X =sale[['Dec-2017','Nov-2017','Oct-2017', 'Sep-2017', 'Aug-2017','Jul-2017']]
y=sale[['CLV']]

Importing required Library

In [45]:
from sklearn.model_selection import train_test_split

<b> Train-test Split </b> 

The train-test split is a technique for evaluating the performance of a machine learning algorithm.

It can be used for classification or regression problems and can be used for any supervised learning algorithm.

The procedure involves taking a dataset and dividing it into two subsets. The first subset is used to fit the model and is referred to as the training dataset. The second subset is not used to train the model; instead, the input element of the dataset is provided to the model, then predictions are made and compared to the expected values. This second dataset is referred to as the test dataset.

*  Train Dataset: Used to fit the machine learning model.
*  Test Dataset: Used to evaluate the fit machine learning model.

In [57]:
X_train,X_test, y_train,y_test= train_test_split(X,y, random_state=101)

### Model Development

Importing the Liner Regression model. Then fit your model on the train set using fit() and then perform prediction on the test set using predict()

In [58]:
from sklearn.linear_model import LinearRegression

In [59]:
# initiate 

linreg = LinearRegression()

# fit the model
linreg.fit(X_train,y_train)

# Predict The model
y_pred= linreg.predict(X_test)

In [60]:
linreg.coef_

array([[0.9903775 , 1.0231265 , 0.95525256, 1.02151772, 0.95859498,
        0.94481856]])

### How well does your model fit the data?

To evaluate how well the overall fit of the linear Regression model, we are going to use the R-Square. It is rthe proportion of variance explained by the model. R-squared value lies between 0 and 1 and higher the value better. 

In [61]:
from sklearn import metrics

In [62]:
print("R-Square:",metrics.r2_score(y_test, y_pred))

R-Square: 0.5412886748967976


This model has R-squared(0.54).This model provides the medium fit to the data. 